corehq/apps/case_importer/do_import.py

```diff
@@ -0,0 +1,272 @@
+from __future__ import absolute_import, unicode_literals
+
+import uuid
+from collections import namedtuple
+
+from couchdbkit.exceptions import ResourceNotFound
+
+from casexml.apps.case.const import CASE_TAG_DATE_OPENED
+from casexml.apps.case.mock import CaseBlock, CaseBlockError
+from soil.progress import set_task_progress
+
+from corehq.apps.export.tasks import add_inferred_export_properties
+from corehq.apps.hqcase.utils import submit_case_blocks
+from corehq.apps.locations.models import SQLLocation
+from corehq.apps.users.models import CouchUser
+from corehq.form_processor.interfaces.dbaccessors import CaseAccessors
+from corehq.toggles import BULK_UPLOAD_DATE_OPENED
+from corehq.util.datadog.utils import case_load_counter
+from corehq.util.soft_assert import soft_assert
+
+from . import util as importer_util
+from .const import ImportErrors, LookupErrors
+
+CASEBLOCK_CHUNKSIZE = 100
+RowAndCase = namedtuple('RowAndCase', ['row', 'case'])
+
+
+def do_import(spreadsheet, config, domain, task=None, chunksize=CASEBLOCK_CHUNKSIZE,
+              record_form_callback=None):
+    match_count = created_count = too_many_matches = num_chunks = 0
+    errors = importer_util.ImportErrorDetail()
+
+    user = CouchUser.get_by_user_id(config.couch_user_id, domain)
+    username = user.username
+    user_id = user._id
+
+    # keep a cache of id lookup successes to help performance
+    id_cache = {}
+    name_cache = {}
+    caseblocks = []
+    ids_seen = set()
+    track_load = case_load_counter("case_importer", domain)
+
+    def _submit_caseblocks(domain, case_type, caseblocks):
+        err = False
+        if caseblocks:
+            try:
+                form, cases = submit_case_blocks(
+                    [cb.case.as_string().decode('utf-8') for cb in caseblocks],
+                    domain,
+                    username,
+                    user_id,
+                    device_id=__name__ + ".do_import",
+                )
+
+                if form.is_error:
+                    errors.add(
+                        error=ImportErrors.ImportErrorMessage,
+                        row_number=form.problem
+                    )
+            except Exception:
+                err = True
+                for row_number, case in caseblocks:
+                    errors.add(
+                        error=ImportErrors.ImportErrorMessage,
+                        row_number=row_number
+                    )
+            else:
+                if record_form_callback:
+                    record_form_callback(form.form_id)
+                properties = set().union(*[set(c.dynamic_case_properties().keys()) for c in cases])
```

Comment:
... or ...
```python
    properties = {p for c in cases for p in c.dynamic_case_properties().keys()}
```
... although it's debatable what's more readable, so maybe just consider this a totally optional nit.

Reply:
^ That is more readable, and is a bit more :snake:

Reply:
Oh, oh, for more :snake: we can drop `keys()`:
```python
    properties = {p for c in cases for p in c.dynamic_case_properties()}
```
:grin:

Reply:
:+1: 
Although I actually don't like the implicit keys thing - makes me forget that I'm dealing with a dict.